src/main/java/rx/internal/operators/BlockingOperatorToIterator.java

```diff
@@ -47,68 +46,82 @@ private BlockingOperatorToIterator() {
      * @return the iterator that could be used to iterate over the elements of the observable.
      */
     public static <T> Iterator<T> toIterator(Observable<? extends T> source) {
-        final BlockingQueue<Notification<? extends T>> notifications = new LinkedBlockingQueue<Notification<? extends T>>();
+        SubscriberIterator<T> subscriber = new SubscriberIterator<T>();
 
         // using subscribe instead of unsafeSubscribe since this is a BlockingObservable "final subscribe"
-        final Subscription subscription = source.materialize().subscribe(new Subscriber<Notification<? extends T>>() {
-            @Override
-            public void onCompleted() {
-                // ignore
-            }
+        source.materialize().subscribe(subscriber);
+        return subscriber;
+    }
 
-            @Override
-            public void onError(Throwable e) {
-                notifications.offer(Notification.<T>createOnError(e));
-            }
+    public static final class SubscriberIterator<T>
+        extends Subscriber<Notification<? extends T>> implements Iterator<T> {
 
-            @Override
-            public void onNext(Notification<? extends T> args) {
-                notifications.offer(args);
-            }
-        });
+        private final BlockingQueue<Notification<? extends T>> notifications;
+        private Notification<? extends T> buf;
 
-        return new Iterator<T>() {
-            private Notification<? extends T> buf;
+        public SubscriberIterator() {
+            this.notifications = new LinkedBlockingQueue<Notification<? extends T>>();
+            this.buf = null;
+        }
 
-            @Override
-            public boolean hasNext() {
-                if (buf == null) {
-                    buf = take();
-                }
-                if (buf.isOnError()) {
-                    throw Exceptions.propagate(buf.getThrowable());
-                }
-                return !buf.isOnCompleted();
+        @Override
+        public void onStart() {
+            request(0);
+        }
+
+        @Override
+        public void onCompleted() {
+            // ignore
+        }
+
+        @Override
+        public void onError(Throwable e) {
+            notifications.offer(Notification.<T>createOnError(e));
+        }
+
+        @Override
+        public void onNext(Notification<? extends T> args) {
+            notifications.offer(args);
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (buf == null) {
+                request(1);
```

Comment:
The problem with this approach is that there can't be any cross-thread pipelining because one thread won't progress unless the other thread does 1 thing. A better approach would be to request in bigger batches and re-request after a certain proportion of the batch has been received.